### PR TITLE
Updating modeling documentation

### PR DESCRIPTION
**Pull request by perrygreenfield**
This PR is primarily to add the contents of changes_for_4.rst in modeling. Though I would also like to see the proper form for the reference to this in CHANGES.rst if someone would helpfully give such an example.

EDIT: Follow up of #8769

### CODE REVIEW

**Comment by pllim** [comment]
Maybe squashing commits is in order? 😬

File: docs/modeling/changes_for_4.rst [hunk 1]
```diff
@@ -1,4 +1,181 @@
...
+make the behavior of parameters in compound models more intuitive,
+many changes have been made internally to the modeling code,
+particularly to compound models and parameters. This page
+summarizes the important changes. More technical details are 
+given at the end, but it is generally not necessary to read those
+unless one wants to understand why some changes were necessary.
+
+- Support for expressions of modeling classes has been removed.
+  Expressions of model class instances are still fully supported.
+  This was done to simplify the implemenation, improve performance,
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: typo: implementation

File: docs/modeling/changes_for_4.rst [hunk 2]
```diff
@@ -1,4 +1,181 @@
...
+
+        m = m1 + m2 + m3 + m4
+        m[:2] # Results in m1 + m2, works
+        m[1:] # Should result in m2 + m3 + m4, does not work
+               # since m1 is part of all subexpressions.
+
+
+
+- Generally, all public methods have remained unchanged. The exception
+  is .n_submodels that used to be a method but now is a property.
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: ``.n_submodels`` (in backticks)

also change that -> which

File: docs/modeling/changes_for_4.rst [hunk 3]
```diff
@@ -1,4 +1,181 @@
...
+
+- Many of the non-public methods have changed, particularly for
+  compound models.
+
+- The _CompoundModelMeta metaclass no longer exists.
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: Put class name in backticks

File: docs/modeling/changes_for_4.rst [hunk 4]
```diff
@@ -1,4 +1,181 @@
...
+
+-----------------
+Technical Details
+-----------------
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: This is not the standard way to do sub-headings, I think except for the top-level one it should just be one line below (not above). Can you check another page to make sure?

File: docs/modeling/changes_for_4.rst [hunk 5]
```diff
@@ -1,4 +1,181 @@
...
+Previously Python descriptors were used to define parameters. The
+drawback of descriptors is that all instances of the models that
+they are defined in share the same parameter descriptor instance.
+This means the parameter cannot hold different values for different
+model instances. The way that this was worked around was to have
+the model hold the actual parameter value (and any other information
+relevant to that particular instance of the model). The actual 
+values of the paramters were held in an array that the model
+held, and that array held all values for all parameters in one
+array. The drawbacks of this approach were that compound models
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: were -> was

File: docs/modeling/changes_for_4.rst [hunk 6]
```diff
@@ -1,4 +1,181 @@
...
+had to create their own array holding values for all the parameters
+of the compound model. As a result, the parameter values in a 
+compound model now are completely disassociated with those in
+the constituent model that made up the compound model. Changes 
+to one were not reflected in the other, often leadin to confusion.
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: typo: leading

File: docs/modeling/changes_for_4.rst [hunk 7]
```diff
@@ -1,4 +1,181 @@
...
+The new implementation does away with the metaclass for
+compound models and correspondingly no longer supports
+expressions of model classes, but only expressions of 
+model instances. Previously the expression tree was a private 
+attribute. Now the compound class is itself an expression
+tree. The mapping of constituent parameters to the
+compound is not automatically done except when certain 
+method or attribute calls are mode, which are not normally 
+done on subexpressions. There may be some special cases
+where it is necessary to force this by calling the map_parameters()
```
Comment (astrofrog) [inline docs/modeling/changes_for_4.rst]: backticks around method name

File: docs/modeling/changes_for_4.rst [hunk 8]
```diff
@@ -1,4 +1,181 @@
-Modeling Changes in astropy v.4.0
-*********************************
+.. include:: links.inc
 
-(placeholder)
\ No newline at end of file
+**************************
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion

```

File: docs/modeling/changes_for_4.rst [hunk 9]
```diff
@@ -1,4 +1,181 @@
...
+
+
+In order to make the internal code simpler, improve performance, and
+make the behavior of parameters in compound models more intuitive,
+many changes have been made internally to the modeling code,
+particularly to compound models and parameters. This page
+summarizes the important changes. More technical details are 
+given at the end, but it is generally not necessary to read those
+unless one wants to understand why some changes were necessary.
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
unless you want to understand why some changes were necessary.
```

File: docs/modeling/changes_for_4.rst (hunk 1 above)
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  This was done to streamline the implementation, improve performance,
```

File: docs/modeling/changes_for_4.rst [hunk 10]
```diff
@@ -1,4 +1,181 @@
...
+.. include:: links.inc
 
-(placeholder)
\ No newline at end of file
+**************************
+Changes to Modeling in 4.0
+**************************
+
+
+In order to make the internal code simpler, improve performance, and
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
In order to make the internal code less complex, improve performance, and
```

File: docs/modeling/changes_for_4.rst [hunk 11]
```diff
@@ -1,4 +1,181 @@
...
+        newmodel = Gaussian1D(3.2, 7.1, 2.1) + Const1D(3.)
+
+- Previous to 4.0, parameters were class descriptors, which meant
+  that they could not hold values for the models. Instead, the 
+  values were held inside the models. This resulted in confusion
+  when compound models were used since this necessitated that
+  the compound models make copies of the values. As a result, 
+  changing the value in the compound model did not change the
+  constituent model's parameter value and visa versa. Now
+  parameters are distinct instances for each use, and they
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  parameters are distinct instances for each use and they
```

File: docs/modeling/changes_for_4.rst [hunk 12]
```diff
@@ -1,4 +1,181 @@
...
+  do now hold the value of the parameter. So compound models
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  do hold the value of the parameter, so compound models
```

File: docs/modeling/changes_for_4.rst [hunk 13]
```diff
@@ -1,4 +1,181 @@
...
+
+        In [1]: g = Gaussian1D([1,1], [1,2], [2,4], n_models=2)
+        In [2]: g.amplitude
+        Out[2]: Parameter('amplitude', value=[1. 1.])
+        In [3]: g.amplitude.shape
+        Out[3]: (2,)
+
+- Previously the values were held in an array within the model
+  instance and it was possible to assign values to slices of
+  that array. Reassigning the array does update the parameters,
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  that array. Reassigning the array did update the parameters,
```

File: docs/modeling/changes_for_4.rst [hunk 14]
```diff
@@ -1,4 +1,181 @@
...
+  but assigning slices does not. The new approach is to either
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  but assigning slices did not. The new approach is to either
```

File: docs/modeling/changes_for_4.rst [hunk 15]
```diff
@@ -1,4 +1,181 @@
...
+  replace the whole array by assigning to the parameters property
+  or assign directly to the parameter value.
+
+- The use of 'imputed' units, i.e., supplying input/output units
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
- The use of 'imputed' units (i.e., supplying input/output units
```

File: docs/modeling/changes_for_4.rst [hunk 16]
```diff
@@ -1,4 +1,181 @@
...
+  to a compound model without them but where the component models 
+  support the _parameter_units_for_data_units() method is much more
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  support the _parameter_units_for_data_units() method) is much more
```

Comment (lglattly) [reply docs/modeling/changes_for_4.rst]: Can you try rephrasing this sentence for clarity? Perhaps putting the i.e. material in its own sentence would help?

File: docs/modeling/changes_for_4.rst [hunk 17]
```diff
@@ -1,4 +1,181 @@
...
+
+  permitted this slice::
+
+        m[1:] # results in m2 + m3
+
+  Now, only submodels in the expression tree (think of it as
+  the sequence of operations as performed) are permitted as 
+  slices. This means some slices that make sense do not work
+  now. The following code illustrates what is permitted 
+  and what isn't::
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  and what is not::
```

File: docs/modeling/changes_for_4.rst [hunk 18]
```diff
@@ -1,4 +1,181 @@
...
+
+        m = m1 + m2 + m3 + m4
+        m[:2] # Results in m1 + m2, works
+        m[1:] # Should result in m2 + m3 + m4, does not work
+               # since m1 is part of all subexpressions.
+
+
+
+- Generally, all public methods have remained unchanged. The exception
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
- Generally, all public methods remain unchanged. The exception
```

File: docs/modeling/changes_for_4.rst (hunk 2 above)
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
  is ``.n_submodels`` which used to be a method but now is a property.
```

File: docs/modeling/changes_for_4.rst (hunk 3 above)
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
- The ``_CompoundModelMeta`` metaclass no longer exists.
```

File: docs/modeling/changes_for_4.rst [hunk 19]
```diff
@@ -1,4 +1,181 @@
...
+  is .n_submodels that used to be a method but now is a property.
+
+- Many of the non-public methods have changed, particularly for
+  compound models.
+
+- The _CompoundModelMeta metaclass no longer exists.
+
+-----------------
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
```

File: docs/modeling/changes_for_4.rst (hunk 4 above)
Comment (lglattly) [reply docs/modeling/changes_for_4.rst]: The standard for headings is just the line below. I'm not sure if the line above actually changes the hierarchy rendering but I suggest taking them out for consistency!

File: docs/modeling/changes_for_4.rst [hunk 20]
```diff
@@ -1,4 +1,181 @@
...
+Technical Details
+-----------------
+
+~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
```

File: docs/modeling/changes_for_4.rst [hunk 21]
```diff
@@ -1,4 +1,181 @@
...
+Parameter-related changes
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
Parameter-Related Changes
```

File: docs/modeling/changes_for_4.rst [hunk 22]
```diff
@@ -1,4 +1,181 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Previously Python descriptors were used to define parameters. The
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
Previously, Python descriptors were used to define parameters. The
```

File: docs/modeling/changes_for_4.rst [hunk 23]
```diff
@@ -1,4 +1,181 @@
...
+
+- The use of 'imputed' units, i.e., supplying input/output units
+  to a compound model without them but where the component models 
+  support the _parameter_units_for_data_units() method is much more
+  restricted in its applicability. That will only work when the
+  compound expression uses the '+' or '-' operators. Past behavior
+  led to sometimes arbitrary assignments of units, and sometimes
+  incorrect units to the parameters.
+
+- Slicing is more restrictive now. Previously a model defined as
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
- Slicing is more restrictive now. Previously, a model defined as 
```

File: docs/modeling/changes_for_4.rst [hunk 24]
```diff
@@ -1,4 +1,181 @@
...
+
+~~~~~~~~~~~~~~~~~~~~~~~~~
+Parameter-related changes
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Previously Python descriptors were used to define parameters. The
+drawback of descriptors is that all instances of the models that
+they are defined in share the same parameter descriptor instance.
+This means the parameter cannot hold different values for different
+model instances. The way that this was worked around was to have
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
model instances. The way users worked around this was to have
```

File: docs/modeling/changes_for_4.rst (hunk 5 above)
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
array. The drawback of this approach was that compound models
```

File: docs/modeling/changes_for_4.rst (hunk 6 above)
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
to one were not reflected in the other, often leading to confusion.
```

File: docs/modeling/changes_for_4.rst [hunk 25]
```diff
@@ -1,4 +1,181 @@
...
+held, and that array held all values for all parameters in one
+array. The drawbacks of this approach were that compound models
+had to create their own array holding values for all the parameters
+of the compound model. As a result, the parameter values in a 
+compound model now are completely disassociated with those in
+the constituent model that made up the compound model. Changes 
+to one were not reflected in the other, often leadin to confusion.
+
+The new implementation still uses parameter descriptors, but only
+as an intermediate solution. In creating an instance of the model
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
as an intermediate solution. In creating an instance of the model,
```

File: docs/modeling/changes_for_4.rst [hunk 26]
```diff
@@ -1,4 +1,181 @@
...
+the descriptor instance is used to create a local instance of a
+parameter that is not shared between model instances. So now
+parameters do hold all the information directly, and can be
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
parameters do hold all of the information directly, and can be
```

File: docs/modeling/changes_for_4.rst [hunk 27]
```diff
@@ -1,4 +1,181 @@
...
+shared between compound and constituent models.
+
+One consequence is that the interface to fitting engines still
+use the model parameter array, but this array is constructed
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
uses the model parameter array, but this array is constructed
```

File: docs/modeling/changes_for_4.rst [hunk 28]
```diff
@@ -1,4 +1,181 @@
...
+from the parameters on the fly. One may set a completely new
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
from the parameters on the fly. You may set a completely new
```

File: docs/modeling/changes_for_4.rst [hunk 29]
```diff
@@ -1,4 +1,181 @@
...
+array; the model is defined to detect such assignments and 
+back propagate the values to the parameters. For example, when
+doing a fit, the fit results are propagated to the parameters
+in such a way. But if one assigns a slice to the model parameter
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
in such a way. But if you assign a slice to the model parameter
```

File: docs/modeling/changes_for_4.rst [hunk 30]
```diff
@@ -1,4 +1,181 @@
...
+array, there is no simple way to detect that assignment and
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
array, there is no convenient way to detect that assignment and
```

File: docs/modeling/changes_for_4.rst [hunk 31]
```diff
@@ -1,4 +1,181 @@
...
+make the necessary parameter updates. So this mode no longer
+works unless an explicit method call is made to back propagate
+the values. It is not recommended to use this kind of interface
+now.
+
+Likewise, parameter-specific attributes also are kept by the
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
Likewise, parameter-specific attributes are also kept by the 
```

File: docs/modeling/changes_for_4.rst [hunk 32]
```diff
@@ -1,4 +1,181 @@
...
+the values. It is not recommended to use this kind of interface
+now.
+
+Likewise, parameter-specific attributes also are kept by the 
+parameter, such as the constraints on the parameter minimum or
+maximum, and whether it is fixed as far as fitting is concerned.
+
+Since the parameter no longer is required to link to a specific
+model, it holds any dimension corresponding to the model_set
+size; when it previously did not.
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
size, when it previously did not.
```

File: docs/modeling/changes_for_4.rst [hunk 33]
```diff
@@ -1,4 +1,181 @@
...
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
```

File: docs/modeling/changes_for_4.rst [hunk 34]
```diff
@@ -1,4 +1,181 @@
...
+Compound Model implementation changes
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
Compound Model Implementation Changes
```

File: docs/modeling/changes_for_4.rst [hunk 35]
```diff
@@ -1,4 +1,181 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Compound models previously were implemented using a metaclass 
+for the compound model while also inheriting from the Model
+class (which itself has a metaclass). The primary reason for
+this approach was to support expressions of Model classes. 
+However, this leads to a very hard to understand implementation,
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
However, this leads to a confusing implementation,
```

File: docs/modeling/changes_for_4.rst [hunk 36]
```diff
@@ -1,4 +1,181 @@
...
+expressions of model classes also include model instances.
+
+The new implementation does away with the metaclass for
+compound models and correspondingly no longer supports
+expressions of model classes, but only expressions of 
+model instances. Previously the expression tree was a private 
+attribute. Now the compound class is itself an expression
+tree. The mapping of constituent parameters to the
+compound is not automatically done except when certain 
+method or attribute calls are mode, which are not normally
```
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
method or attribute calls are made, which are not normally 
```

File: docs/modeling/changes_for_4.rst (hunk 7 above)
Comment (lglattly) [inline docs/modeling/changes_for_4.rst]:
```suggestion
where it is necessary to force this by calling the ``map_parameters()``
```

**Comment by lglattly** [comment]
I reviewed and made some copy editing suggestions, let me know if there are any questions!

File: docs/modeling/changes_for_4.rst (hunk 5 above)
Comment (perrygreenfield) [reply docs/modeling/changes_for_4.rst]: done

**Comment by perrygreenfield** [comment]
being superseded by a new PR #9077

**Pull request closed without merging**